proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,182 @@
+
+### Audience
+
+* Prometheus maintainers
+* Alertmanager maintainers
+* Secret providers interested in contributing code
+* Users looking to use secret providers
+
+## Non-Goals
+
+* Implement a variety of secret providers.
+
+## How
+
+
+### Interfaces
+
+Secret providers will be created from their configurations through the following interface:
+
+```
+type SecretProviderConfiguration interface {
+	// Returns the secret provider for the given configuration.
+	Load() (SecretProvider, error)
+}
+```
+
+Secret providers will have to satisfy the following interface. Secret providers will be expected to be long lived to allow for caching. However they will be re-instantiated in case of a configuration change. The Fetch method will be called before each http request done through `prometheus/common/config`
+
+```
+type SecretProvider interface {
+	// Returns the secret value for the given configuration.
+  Fetch(ctx context.Context, secretRef string) (string, error)
+}
+```
+
+### Configuration
+
+Globally there will be a section to configure secrete providers to be used throught the config file. Here is an example:
+
+```
+global:
+  secret_providers:
+  - name: my_secret_provider
+    kubernetes_sp_config:
+      namespace: ns1
+```
+
+
+For secret related fields under http_config, a new `ref` variant will be added that can reference these secret providers. For instance, basic_auth will have the following form
+
+```
+basic_auth:
+...
+  password: <secret>
+  password_file: <string>
+  password_ref:
+    <string>: <string>
+```
+
+and could be instantiated as follows:
+
+```
+basic_auth:
+  password_ref:
+    my_secret_provider: 'my-secret-key'
+```
+
+#### Full configuration example
+```
+global:
+  secret_providers:
+  - name: kube1
+    kubernetes_sp_config:
+      namespace: ns1
+  - name: kube2
+    kubernetes_sp_config:
+      namespace: ns2
+...
+  scrape_configs:
+  - job_name: 'http-basic-auth-endpoint'
+    http_config:
+      basic_auth:
+        username: 'myuser'
+        password_ref:
+          kube1: 'myuser-pass'
+    static_configs:
+    - targets: ['www.endpoint.com/basic-auth']
+  scrape_configs:
+  - job_name: 'http-authorization-auth-endpoint'
+    http_config:
+      authorization:
+        credentials_ref:
+          kube2: 'header-credentials'
+    static_configs:
+    - targets: ['www.endpoint.com/authorization-auth']
+  scrape_configs:
+  - job_name: 'tls-certificate-endpoint'
+    http_config:
+      tls_config:
+        key_ref:
+          kube2: 'header-credentials'
+    static_configs:
+    - targets: ['www.endpoint.com/tls-certificate']
+```
+
+
+## Alternatives
+
+### Modify Secret type
+
+Currently most secrets in the config use the prometheus.common.config.Secret alias. We could modify this type such that if only a string is passed it behaves the same as before, and if a map is passed in it assumes it to be a reference to be fetched from the associated secret provider.
```

Comment:
Sounds like this alternative proposes a way to reuse `password` key (instead of adding new `password_ref`). I think this is very good to discuss and decide what's better 👍🏽 

(I would make this explicit and reduce mentions of code interfaces for now)

Reply:
I like it too, but we have to be careful -- a loosely typed field like that can't (nicely) be represented in some schemas such as a Kubernetes CRD. I got strong pushback in the past when I proposed using something like this in just the unit test format.

Reply:
Moved it around to be the main proposal and switched to suggest yaml tags